### PR TITLE
p2p: Add p2p module framework

### PR DESCRIPTION
**Pull request by qiaozhigang**
1. create log.GetLogger function，can create each logger by name
2. move p2p/network to p2p/qvic, and modify package name to qvic
3. add p2p/simulations folder, put complex test routine.
4. p2p/simulations/p2ptest.go shows how to user p2p module in a higher level

### CODE REVIEW

File: log/log.go [hunk 1]
```diff
@@ -79,3 +82,26 @@ func init() {
 	log.Out = os.Stdout
 	log.SetLevel(logrus.DebugLevel)
 }
+
+// GetLogger get logrus.Logger object accoring to logName
+// each module can have it's own logger
+func GetLogger(logName string, bConsole bool) *logrus.Logger {
+	curLog, ok := logMap[logName]
```
Comment (lucusfly) [inline log/log.go]: we can't return *logrus.Logger to external, because we may change log implementation later

File: log/log.go [hunk 2]
```diff
@@ -16,6 +17,8 @@ import (
 
 var log *logrus.Logger
 
+var logMap map[string]*logrus.Logger
```
Comment (lucusfly) [inline log/log.go]: I didn't find where you set logMap

File: p2p/discovery/server.go [hunk 3]
```diff
@@ -27,6 +27,18 @@ func StartServer(port, id string) {
 	wg := sync.WaitGroup{}
 	wg.Add(1)
 	wg.Wait()
+
+}
+
```
Comment (lucusfly) [inline p2p/discovery/server.go]: you'd better to resolve the old PR comments and send a new iteration

File: vendor/vendor.json [hunk 4]
```diff
@@ -4,233 +4,231 @@
 	"package": [
 		{
 			"checksumSHA1": "YdWmkLgfIqoZgQA0bWB71jCizpc=",
+			"origin": "github.com/seeleteam/vendor/github.com/aristanetworks/goarista/monotime",
 			"path": "github.com/aristanetworks/goarista/monotime",
-			"revision": "bdff49616af24181f3b83535ed73127c615ba55a",
-			"revisionTime": "2018-01-24T01:13:10Z"
+			"revision": ""
 		},
```
[the remark this answers is not in the document]
Comment (qiaozhigang) [reply vendor/vendor.json]: modified by govendor automatically

File: log/log.go (hunk 1 above)
Comment (qiaozhigang) [reply log/log.go]: So, need create a log wrapper.
Delete all logs at this time

File: log/log.go [hunk 5]
```diff
@@ -7,15 +7,57 @@ package log
 
 import (
 	//"fmt"
```
Comment (WTer) [inline log/log.go]: remove this line.

File: log/log.go [hunk 6]
```diff
@@ -7,15 +7,57 @@ package log
...
+	"fmt"
 	"os"
 
 	"github.com/sirupsen/logrus"
 	//"os/user"
 	//"path/filepath"
```
Comment (WTer) [inline log/log.go]: remove these 2 lines.

File: log/log.go [hunk 7]
```diff
@@ -7,15 +7,57 @@ package log
...
 )
 
+//SeeleLog wrapped log class
```
Comment (WTer) [inline log/log.go]: Add a blank between '//' and 'SeeleLog'

File: p2p/discovery/server.go [hunk 8]
```diff
@@ -29,6 +29,17 @@ func StartService(port, id string) {
 	wg := sync.WaitGroup{}
 	wg.Add(1)
 	wg.Wait()
+
```
Comment (WTer) [inline p2p/discovery/server.go]: remove this blank line.

File: p2p/peer.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+const (
+	pingInterval         = 3 * time.Second // ping interval for peer tcp connection. Should be 15
+	discAlreadyConnected = 10              // node already has connection
+	discServerQuit       = 11              // p2p.server need quit, all peers should quit as it can
+)
+
+// Peer represents a connected remote node.
+type Peer struct {
+	fd       net.Conn        // tcp connection
```
Comment (WTer) [inline p2p/peer.go]: it would be better to rename fd to conn.

File: p2p/protocol.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	Name string
+
+	// Version should contain the version number of the protocol.
+	Version uint
+	// AddPeerCh a peer joins protocol, SubProtocol should handle the channel
+	AddPeerCh chan *Peer
+
+	// DelPeerCh a peer leaves protocol
+	DelPeerCh chan *Peer
+
```
Comment (lucusfly) [inline p2p/protocol.go]: rename to DeletePeerCh

File: p2p/protocol.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+func (p *Protocol) cap() Cap {
+	return Cap{p.Name, p.Version}
+}
+
+// Cap is the structure of a peer capability.
+type Cap struct {
+	Name    string
+	Version uint
+}
+
```
Comment (lucusfly) [inline p2p/protocol.go]: why do you call it capability

File: p2p/server.go [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/aristanetworks/goarista/monotime"
+	"github.com/seeleteam/go-seele/common"
+	"github.com/seeleteam/go-seele/log"
+	"github.com/seeleteam/go-seele/p2p/discovery"
+	//"github.com/ethereum/go-ethereum/p2p/discover"
```
Comment (lucusfly) [inline p2p/server.go]: remove

File: p2p/server.go [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+	lock    sync.Mutex // protects running
+	running bool
+
+	kadDB    *discovery.Database
+	listener net.Listener
+
+	quit chan struct{}
+
+	addpeer chan *Peer
+	delpeer chan *Peer
```
Comment (lucusfly) [inline p2p/server.go]: deletePeer

File: log/log.go [hunk 14]
```diff
@@ -79,3 +121,48 @@ func init() {
...
+}
+
+// GetLogger get logrus.Logger object accoring to logName
+// each module can have it's own logger
+func GetLogger(logName string, bConsole bool) *SeeleLog {
+	if logMap == nil {
+		logMap = make(map[string]*SeeleLog)
+	}
+	curLog, ok := logMap[logName]
+	if ok {
```
[the remark this answers is not in the document]
Comment (qiaozhigang) [reply log/log.go]: add mutex

**Pull request merged**